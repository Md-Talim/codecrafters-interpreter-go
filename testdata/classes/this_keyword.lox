// Test Case 1
class Calculator {
  add(a, b) {
    return a + b + this.memory;
  }
}

var calc = Calculator();
calc.memory = 82;
print calc.add(92, 1);

// Test Case 2
class Animal {
  makeSound() {
    print this.sound;
  }
  identify() {
    print this.species;
  }
}

var dog = Animal();
dog.sound = "Woof";
dog.species = "Dog";

var cat = Animal();
cat.sound = "Meow";
cat.species = "Cat";

// Swap methods between instances
cat.makeSound = dog.makeSound;
dog.identify = cat.identify;

cat.makeSound();
dog.identify();

// Test Case 3
class Wizard {
  getSpellCaster() {
    fun castSpell() {
      print this;
      print "Casting spell as " + this.name;
    }

    return castSpell;
  }
}

var wizard = Wizard();
wizard.name = "Merlin";
wizard.getSpellCaster()();