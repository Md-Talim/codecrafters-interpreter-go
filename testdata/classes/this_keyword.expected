175
Woof
Cat
Wizard instance
Casting spell as Merlin